include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1094 @@
+    TapedDrivers<ADNumberType,ScalarType,T>::jacobian (
+      const unsigned int &,
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcNotImplemented());
+    }
+
+
+    // Specialization for taped Adol-C auto-differentiable numbers.
+    //
+    // Note: In the case of Adol-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template<typename ADNumberType>
+    struct TapedDrivers<ADNumberType,double,typename std::enable_if<
+      ADNumberTraits<ADNumberType>::type_code == NumberTypes::adolc_taped
+      >::type>
+    {
+      typedef double scalar_type;
+
+      // === Taping ===
+
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep)
+      {
+        trace_on(tape_index,keep);
+      }
+
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file)
+      {
+        if (write_tapes_to_file)
+          {
+            trace_off(active_tape_index); // Slow
+
+            std::vector<std::size_t> counts (STAT_SIZE);
+#ifdef DEAL_II_WITH_ADOLC
+            ::tapestats(active_tape_index, counts.data());
+#else
+            AssertThrow(false, ExcNotImplemented());
+#endif
+          }
+        else
+          trace_off(); // Fast(er)
+      }
+
+      // === Scalar drivers ===
+
+      static scalar_type
+      value (const unsigned int             &active_tape_index,
+             const unsigned int             &n_independent_variables,
+             const std::vector<scalar_type> &independent_variables)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+
+        scalar_type value = 0.0;
+
+#ifdef DEAL_II_WITH_ADOLC
+        ::function(active_tape_index,
+                   1, // Only one dependent variable
+                   n_independent_variables,
+                   const_cast<double *>(independent_variables.data()),
+                   &value);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        return value;
+      }
+
+      static void
+      gradient (const unsigned int             &active_tape_index,
+                const unsigned int             &n_independent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                Vector<scalar_type>            &gradient)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(gradient.size() == n_independent_variables,
+               ExcMessage("The length of the gradient vector must be equal "
```

Comment:
i would just do `ExcDimensionMismatch` as it will give you the actual numbers as opposed to just a message. 

Same below in numerous places.

Reply:
Done.